### PR TITLE
Exporter tuning

### LINKED ISSUE #686: Sort questionnaires in exporter
In the exported Excel file the questionnaires should be sorted in the same order as on the questionnaire overview page.

### PR DESCRIPTION
**Pull request by janno42**
- This sorts the questionnaires in the export file (fix #686) and removes all empty results and TextResults.
- It adds an option to include unpublished courses in the export.
- It also fixes the layout of the exporter view and cell borders in the Excel file.

### CODE REVIEW

File: evap/evaluation/models.py [hunk 1]
```diff
@@ -114,13 +114,27 @@ class Questionnaire(models.Model, metaclass=LocalizeModelBase):
...
-        ordering = ('index', 'name_de')
+        ordering = ('is_for_contributors', 'index', 'name_de')
         verbose_name = _("questionnaire")
         verbose_name_plural = _("questionnaires")
 
     def __str__(self):
         return self.name
 
+    def __lt__(self, other):
+        if self.is_for_contributors and not other.is_for_contributors:
```
Comment (karyon) [inline evap/evaluation/models.py]: `return (self.is_for_contributors, self.index) < (other.is_for_contributors, other.index)`

File: evap/evaluation/tools.py [hunk 2]
```diff
@@ -177,21 +179,25 @@ def _calculate_results_impl(course):
...
+                    continue
                 allowed_states = [TextAnswer.PRIVATE, TextAnswer.PUBLISHED]
                 answers = get_textanswers(contribution, question, allowed_states)
                 results.append(TextResult(question=question, answers=answers))
 
         section_warning = questionnaire_max_answers[(questionnaire, contribution)] < questionnaire_warning_thresholds[questionnaire]
 
-        sections.append(ResultSection(questionnaire, contribution.contributor, contribution.label, results, section_warning))
+        if not export or results:
+            sections.append(ResultSection(questionnaire, contribution.contributor, contribution.label, results, section_warning))
```
Comment (karyon) [inline evap/evaluation/tools.py]: i would prefer skipping text questions and questions with no answers in the exporter. i'm wondering whether there are any code changes required at all? text questions seem to be skipped already.

File: evap/results/exporters.py [hunk 3]
```diff
@@ -60,15 +61,20 @@ def init_styles(self, workbook):
             palette_index = self.CUSTOM_COLOR_START + self.NUM_GRADE_COLORS + i
             style_name = self.deviation_to_style(deviation)
             self.create_style(workbook, deviation_base_style, style_name, palette_index, color)
+            self.create_style(workbook, deviation_base_style + ', left medium', style_name + '_total', palette_index, color)
```
Comment (karyon) [inline evap/results/exporters.py]: have you checked whether it's possible at all to have so many styles? at least on colors there is a surprisingly low maximum number.

also, i have the feeling that in create_style, there are now colors overwritten, please check that.

File: evap/results/exporters.py [hunk 4]
```diff
@@ -60,15 +61,20 @@ def init_styles(self, workbook):
...
 
 
-    def grade_to_style(self, grade):
+    def grade_to_style(self, grade, total=False):
+        if total:
+            return 'grade_' + str(self.normalize_number(grade)) + "_total"
         return 'grade_' + str(self.normalize_number(grade))
```
Comment (karyon) [inline evap/results/exporters.py]: please avoid the code duplication, e.g.
```
style_name = 'grade_' + str(self.normalize_number(grade))
if total:
    style_name += "_total"
return style_name
```

File: evap/results/exporters.py (hunk 3 above)
Comment (janno42) [inline evap/results/exporters.py]: 4.000 styles are allowed per document. so we should be fine.
i will change the color definitions.

File: evap/evaluation/models.py (hunk 1 above)
Comment (janno42) [inline evap/evaluation/models.py]: check

File: evap/evaluation/tools.py (hunk 2 above)
Comment (janno42) [inline evap/evaluation/tools.py]: check

File: evap/results/exporters.py (hunk 4 above)
Comment (janno42) [inline evap/results/exporters.py]: check

**Comment by janno42** [comment]
updated

File: evap/results/exporters.py [hunk 5]
```diff
@@ -185,11 +192,13 @@ def writen(exporter, label="", style_name="default"):
     exporter.row += 1
     writec(exporter, label, style_name)
 
+
 def writec(exporter, label, style_name, rows=1, cols=1):
     """Write the cell in the next column of the current line."""
     _write(exporter, label, exporter.styles[style_name], rows, cols)
     exporter.col += 1
 
+
```
Comment (karyon) [inline evap/results/exporters.py]: it seems like we're using one blank line between methods in most other places, so plz remove this

**Comment by karyon** [comment]
:walking:

File: evap/results/exporters.py (hunk 5 above)
Comment (janno42) [inline evap/results/exporters.py]: methods in classes. i think we do have two newlines outside. https://www.python.org/dev/peps/pep-0008/#blank-lines

**Pull request merged**